documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<JoystickState, float>>? AxisMove;
+    public event Action<AxisEvent<JoystickState, Vector2>>? HatMove;
+}
+```
+
+Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(in TDevice DeviceState, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(in TDevice DeviceState, string? OldText, string? NewText);
+```
+
+This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
+
+`SetPosition` allows moving the mouse cursor without the end user physically moving their mouse.
+
+The device state returned by `State` fills out the following structure:
+
+```cs
+public struct MouseState
+{
+    public IMouse Device { get; init; }
+    public MouseButtonState Buttons { get; init; }
+    public ICursorConfiguration Cursor { get; } // forwards to Device.Cursor
+    public Vector2 Position { get; set; } // all sets after the first set forward to Device.SetPosition
```

Comment:
Why is this settable, I get that the user can move this - but this won't mutate the state!
I think there should instead be a method on the device to do this, and this should stay init-only

Reply:
This is settable for ease of use, forwarding to SetPosition. The average end user doesn't care about our reasoning that this is immutable state and to mutate it you need to go through another property. They just want to write their code in a clean way, so these forwarding properties enable that.

However, if you think that this state structure being readonly (and thus `Position` not updating after `set`), perhaps we should remove the `readonly` restriction. This API is focused on being easy-to-use, and the immutable state is just a courtesy for the Working Group's feedback of wanting to be able to capture a state snapshot with ease.

Reply:
Hmm, I think the end users code will actually end up better and more obvious if we stick to state being 100% immutable. I think this for two reasons:
- The user can end up setting this value and reading it back will result in seemingly no change
- When trying to architect larger solutions involving this code passing the state should do just that - pass the state. Not give full control over the mouse/device to the called method.
I think this ultimately is the same discussion as having the backwards reference above.

Reply:
But the real question is, how to we expose the device then? The input context today doesn't expose the devices any other way.

We'd have to essentially have the lists iterate a `(IMouse Device, MouseState State)` which at that point they are just as tied as they would be if `IMouse` is in here.

No matter what we do, someone somewhere will complain when they set the position, either via `Device.SetPosition`, the original `Position` `set`ter, whatever; that the state is not updated. I'm leaning more towards just implicitly updating the state if the user explicitly changes it to alleviate confusion.

Reply:
> I'm leaning more towards just implicitly updating the state
I'd be strongly against this - the whole point is that the state represents a snapshot in time.

> We'd have to essentially have the lists iterate a (IMouse Device, MouseState State)
Why? It's just fine if it iterates `IMouse`, from where the state can be retrieved using `mouse.State`?
I don't think removing the state interacting with the device changes anything about the rest of the design?

Reply:
That's fair, you've worn me down. Will adjust.

Reply:
Oh but are you fine with me changing the state properties to `ref readonly MouseState` so users don't accidentally copy lots (as they inevitably would e.g. `if (mouse.State.Buttons[MouseButton.LeftButton] || mouse.State.Buttons[MouseButton.MiddleButton)` etc)

Reply:
Resolved - input context now enumerates devices.